types/index.d.ts

```diff
@@ -4,95 +4,194 @@
 
-export type CodecTransformer = (buffer: Buffer, callback: () => void) => Buffer; // TODO
+type Codec = (buffer: Buffer, callback: Callback<Buffer>) => void;
 
-export interface CodecOptions {
-  deflate: CodecTransformer;
-  snappy: CodecTransformer;
+interface CodecOptions {
+  [name: string]: Codec;
 }
 
-export interface Decoder {
-  on(type: 'metadata', callback: (type: Type) => void): this;
-  on(type: 'data', callback: (value: object) => void): this;
+interface DecoderOptions {
+  noDecode: boolean;
+  readerSchema: string | object | Type;
+  codecs: CodecOptions;
+  parseHook: (schema: Schema) => Type
 }
 
-export interface Encoder {
-  // TODO
+interface EncoderOptions {
+  blockSize: number;
+  codec: string;
+  codecs: CodecOptions;
+  writeHeader: boolean | 'always' | 'never' | 'auto';
+  syncMarker: Buffer;
 }
 
-export interface ReaderOptions {
-  // TODO
+interface ForSchemaOptions {
+  assertLogicalTypes: boolean;
+  logicalTypes: { [type: string]: types.LogicalType };
+  namespace: string;
+  noAnonymousTypes: boolean;
+  registry: { [name: string]: Type };
+  typeHook: (schema: Schema, opts: ForSchemaOptions) => Type;
+  wrapUnions: boolean | 'auto' | 'always' | 'never';
 }
 
+interface TypeOptions extends ForSchemaOptions {
+  strictDefaults: boolean;
+}
+
+interface ForValueOptions extends TypeOptions {
+  emptyArrayType: Type;
+  valueHook: (val: any, opts: ForValueOptions) => Type;
+}
+
+interface CloneOptions {
+  coerceBuffers: boolean;
+  fieldHook: (field: types.Field, value: any, type: Type) => any;
+  qualifyNames: boolean;
+  skipMissingFields: boolean;
+  wrapUnions: boolean;
+}
+interface IsValidOptions {
+  noUndeclaredFields: boolean;
+  errorHook: (path: string[], val: any, type: Type) => void
+}
 interface AssembleOptions {
   importHook: (filePath: string, type: 'idl', callback: Callback<object>) => void;
 }
 
-export function assemble(args: any): any;
+interface SchemaOptions {
+  exportAttrs: boolean;
+  noDeref: boolean;
+}
+
+declare class Resolver {
+  //no public methods
+}
+
+//exported functions
+
 export function assembleProtocol(filePath: string, opts: Partial<AssembleOptions>, callback: Callback<object>): void;
 export function assembleProtocol(filePath: string, callback: Callback<object>): void;
-export function combine(args: any): any;
-export function createFileDecoder(fileName: string, codecs?: Partial<CodecOptions>): Decoder;
-export function createFileEncoder(filePath: string, schema: Schema, options?: any): Encoder;
+export function createFileDecoder(fileName: string, opts?: Partial<DecoderOptions>): streams.BlockDecoder;
+export function createFileEncoder(filePath: string, schema: Schema, opts?: Partial<EncoderOptions>): streams.BlockEncoder;
+export function createBlobEncoder(schema: Schema, opts?: Partial<EncoderOptions>): stream.Duplex;
+export function createBlobDecoder(blob: Blob, opts?: Partial<DecoderOptions>): streams.BlockDecoder;
```

Comment:
FWIW I had to add `dom` as a `lib` to my `tsconfig.json` for typescript to compile O.K. after this change.
Not sure if this is something we can work around though in this file. @reuzel 

    "lib": ["dom"]

Reply:
I think a triple slash directive should make it work, but it's not taking effect on my local
    `/// <reference types="dom" />`

Reply:
The triple slash syntax is rather outdated. I never use it anymore. In this case, the lib option seems the right way forward. It is just telling the typescript compiler that the dom objects are assumed to be available. Without, the Blob type would be unknown.

Why do you see this as an issue? As far as I know, the lib option is not transitive, in the sense that clients of your library do not need to add them as long as they do not include the Blob encoders/decoders. As you have nicely splitted that functionality into a browser lib, I don't really see an issue?

By the way, I didn't encounter this issue in my own project. I'm using webpack for the client pieces, and when set in the right mode, WebPack enables the dom objects automatically.